plenum/server/replica.py

```diff
@@ -614,7 +614,7 @@ def primaryName(self, value: Optional[str]) -> None:
                 # decided.
                 return
             self._gc_before_new_view()
-            if self.viewNo > 0:
+            if self.viewNo > 0 and self.last_ordered_3pc[1] == 0:
```

Comment:
I think `self.last_ordered_3pc[1] == 0` is not enough. `last_ordered_3pc` can be from the previous view, and in this case we do need to reset the watermarks. 
A correct condition should be:
```
def __should_reset_watermarks_before_new_view()
    if self.viewNo <= 0:
        return False
    if self.last_ordered_3pc[0] == self.viewNo and self.last_ordered_3pc[1] > 0:
        return False
    return True
```